File: src/main/java/app/askresume/oauth/service/SocialLoginApiServiceFactory.java
package app.askresume.oauth.service;

import app.askresume.domain.member.constant.MemberType;
import org.springframework.stereotype.Service;

import java.util.Map;

@Service
public class SocialLoginApiServiceFactory {

    private static Map<String, SocialLoginApiService> socialLoginApiServices;

    public SocialLoginApiServiceFactory(Map<String, SocialLoginApiService> socialLoginApiServices) {
        this.socialLoginApiServices = socialLoginApiServices;
    }

    public static SocialLoginApiService getSocialLoginApiService(MemberType memberType) {
        String socialLoginApiServiceBeanName = "";

        if(MemberType.GOOGLE.equals(memberType)) {
            socialLoginApiServiceBeanName = "googleLoginApiServiceImpl";
        }
        return socialLoginApiServices.get(socialLoginApiServiceBeanName);
    }

}

File: src/main/kotlin/app/askresume/oauth/service/SocialLoginApiServiceFactory.kt
package app.askresume.oauth.service

import app.askresume.domain.member.constant.MemberType
import app.askresume.global.error.ErrorCode
import app.askresume.global.error.exception.BusinessException
import org.springframework.stereotype.Service

@Service
class SocialLoginApiServiceFactory(
    private val socialLoginApiServices: Map<String, SocialLoginApiService>
) {

    fun getSocialLoginApiService(memberType: MemberType): SocialLoginApiService {
        var socialLoginApiServiceBeanName = ""

        if (MemberType.GOOGLE == memberType) {
            socialLoginApiServiceBeanName = "googleLoginApiServiceImpl"
        }

        return socialLoginApiServices[socialLoginApiServiceBeanName]
            ?: throw BusinessException(ErrorCode.NOT_EXISTS_SOCIAL_LOGIN_API_SERVICE)
    }

}

